### PR TITLE
Add HashHunter hash() update for performance

### PR DESCRIPTION
**Pull request by mgreen27**
Added update to take advantage of hash() algorithm targeting. 
The hash() function will only process hash types that have been requested a match.
Added OSPath notation.

### CODE REVIEW

File: artifacts/definitions/Generic/Detection/HashHunter.yaml [hunk 1]
```diff
@@ -45,14 +45,28 @@ parameters:
...
-      LET SHA256Array <= split(sep='\\s+',string=SHA256List)
+      LET MD5List <= if(condition= MD5List,
+                        then= split(sep='\\s+',string=MD5List), else=Null)
+      LET SHA1List <= if(condition= SHA1List,
+                        then= split(sep='\\s+',string=SHA1List), else=Null)
+      LET SHA256List <= if(condition= SHA256List,
+                        then= split(sep='\\s+',string=SHA256List), else=Null)
+      
+      -- set hash selector for optimised hash calculation
+      LET HashSelector <= split(string=
```
Comment (scudette) [inline artifacts/definitions/Generic/Detection/HashHunter.yaml:56]: How about this

```sql
LET _HashSelector = SELECT * FROM chain(
a={SELECT "MD5" AS Hash FROM scope() WHERE MD5List},
b={SELECT "SHA1" AS Hash FROM scope() WHERE SHA1List},
c={SELECT "SHA256" AS Hash FROM scope() WHERE SHA256List}
)

LET HashSelector <= _HashSelector.Hash
```

Comment (mgreen27) [reply artifacts/definitions/Generic/Detection/HashHunter.yaml]: Thats much cleaner. Just updated.

File: artifacts/definitions/Generic/Detection/HashHunter.yaml [hunk 2]
```diff
@@ -45,14 +45,23 @@ parameters:
...
+      LET SHA1List <= if(condition= SHA1List,
+                        then= split(sep='\\s+',string=SHA1List), else=Null)
+      LET SHA256List <= if(condition= SHA256List,
+                        then= split(sep='\\s+',string=SHA256List), else=Null)
+      
+      -- set hash selector for optimised hash calculation
+      LET HashSelector <= SELECT * FROM chain(
+            a={SELECT "MD5" AS Hash FROM scope() WHERE MD5List},
+            b={SELECT "SHA1" AS Hash FROM scope() WHERE SHA1List},
+            c={SELECT "SHA256" AS Hash FROM scope() WHERE SHA256List})
```
Comment (scudette) [inline artifacts/definitions/Generic/Detection/HashHunter.yaml:59]: probably minor but I was suggesting to pre-calculate this so we dont have to run this query for each file.

File: artifacts/definitions/Generic/Detection/HashHunter.yaml [hunk 3]
```diff
@@ -61,35 +70,35 @@ sources:
...
-         OR if(condition= SHA1List,then= Hash.SHA1 in SHA1Array)
-         OR if(condition= SHA256List, then= Hash.SHA256 in SHA256Array)
-        )
\ No newline at end of file
+      SELECT OSPath,Name,Size,
+            dict(Mtime=Mtime,Atime=Atime,Ctime=Ctime,Btime=Btime) as Timestamps,
+            hash(path=OSPath,hashselect=HashSelector.Hash) as Hash
+        FROM if(condition= HashSelector.Hash, then= find_files)
+        WHERE 
+            (   if(condition= MD5List,then= Hash.MD5 in MD5List)
```
Comment (scudette) [inline artifacts/definitions/Generic/Detection/HashHunter.yaml:101]: you probably do not need the if condition here at all.

File: artifacts/definitions/Generic/Detection/HashHunter.yaml (hunk 2 above)
Comment (mgreen27) [reply artifacts/definitions/Generic/Detection/HashHunter.yaml]: shouldnt LET HashSelector <= SELECT precalculate? I initially left it off then added after realising :)

**Pull request merged**